From 88598d7d37ed230e653b43823f6dc005e77088f3 Mon Sep 17 00:00:00 2001
From: Samreen Nilofer <samreen.nilofer@intel.com>
Date: Fri, 20 Feb 2015 12:02:27 +0530
Subject: [PATCH 06/24] AsoC: soc-fw: Add a check for verifying the plugin
 version

Issue: GMINL-3680

Add a check to verify if the the plugin version is correct.

Change-Id: I842693213553a636b62c5d134c58ab2ab28e3097
Signed-off-by: Samreen Nilofer <samreen.nilofer@intel.com>
---
 include/sound/soc-fw.h |  3 +++
 sound/soc/soc-fw.c     | 15 +++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/include/sound/soc-fw.h b/include/sound/soc-fw.h
index 2f74074..da57b5d 100644
--- a/include/sound/soc-fw.h
+++ b/include/sound/soc-fw.h
@@ -94,6 +94,9 @@ struct snd_soc_fw_platform_ops {
 	/* completion - called at completion of firmware loading */
 	void (*complete)(struct snd_soc_platform *);
 
+	/* Version check */
+	int (*version_check)(u32 version);
+
 	/* kcontrols operations */
 	const struct snd_soc_fw_kcontrol_ops *io_ops;
 	int io_ops_count;
diff --git a/sound/soc/soc-fw.c b/sound/soc/soc-fw.c
index a934758..13a1546 100644
--- a/sound/soc/soc-fw.c
+++ b/sound/soc/soc-fw.c
@@ -420,6 +420,14 @@ static int soc_fw_init_pvt_data(struct soc_fw *sfw, u32 io_type, unsigned long s
 	return 0;
 }
 
+static int soc_verify_plugin_version(struct soc_fw *sfw, u32 version)
+{
+	if (sfw->platform && sfw->platform_ops && sfw->platform_ops->version_check)
+		return sfw->platform_ops->version_check(version);
+
+	return 0;
+}
+
 static int soc_fw_create_tlv(struct soc_fw *sfw, struct snd_kcontrol_new *kc,
 	u32 tlv_size)
 {
@@ -1423,6 +1431,13 @@ static int soc_valid_header(struct soc_fw *sfw, struct snd_soc_fw_hdr *hdr)
 		return -EINVAL;
 	}
 
+	/* Verify the plugin version */
+	if (soc_verify_plugin_version(sfw, hdr->version)) {
+		dev_err(sfw->dev, "ASoC: %s plugin version mismatch hdr version 0x%x\n",
+			hdr->version);
+		return -EINVAL;
+	}
+
 	if (sfw->pass == hdr->type)
 		dev_dbg(sfw->dev,
 			"ASoC: Got 0x%x bytes of type %d version %d vendor %d at pass %d\n",
-- 
1.9.1

